### ISSUE TITLE
Can't use object from a dynamic assembly as global object for script

### ISSUE THREAD

**Issue by daveaglick** [dotnet/roslyn#2246]
To reproduce, create a dynamic assembly using Roslyn:

```
SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(@"
    public class MyGlobals
    {
        ...
    }
");
CSharpCompilation compilation = CSharpCompilation.Create("DynamicAssembly")
    .WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary))
    .AddSyntaxTrees(syntaxTree)
    .AddReferences(...);
byte[] assemblyBytes;
using (MemoryStream ms = new MemoryStream())
{
    compilation.Emit(ms);
    assemblyBytes = ms.ToArray();
}
Assembly assembly = Assembly.Load(assemblyBytes);
Type globalsType = assembly.GetType("MyGlobals");
object globals = Activator.CreateInstance(globalsType);
```

Then try and use the dynamic globals object in a script:

```
ScriptOptions options = new ScriptOptions()
    .AddNamespaces(...)
    .AddReferences(...);
CSharpScript.Eval(@"...", options, globals);
```

It throws an exception `System.ArgumentException : Empty path name is not legal.`. This is the same exception you see if you attempt to call `MetadataReference.CreateFromAssembly()` with a dynamic assembly. Since dynamic assemblies don't have a `Location`, the following call in `MetadataReference.cs" fails:

```
string location = AssemblyLocationLightUp.GetAssemblyLocation(assembly);
Stream peStream = FileStreamLightUp.OpenFileStream(location);
```

Normally I get around this in a compilation by creating my `MetadataReference` from a `Stream`:

```
MetadataReference dynamicReference;
using (MemoryStream ms = new MemoryStream(assemblyBytes))
{
    dynamicReference = MetadataReference.CreateFromStream(ms);
}
```

However, this doesn't work for script globals because of the following line in `Core\Script.cs` (https://github.com/dotnet/roslyn/blob/master/src/Scripting/Core/Script.cs#L375):

```
var globalsTypeAssembly = MetadataReference.CreateFromAssembly(this.GlobalsType.Assembly);
```

**TL;DR: A script always tries to use `MetadataReference.CreateFromAssembly()` to get the assembly metadata for the global object, and that approach doesn't work for global objects defined in dynamic assemblies.**

Not sure what a good fix would be. Perhaps allowing the specification of a `MetadataReference` for the global object to be used? Or checking the provided assembly references for one that contains the global object *before* attempting to get one? Alternatively, any fix for dynamic assemblies causing an exception with `MetadataReference.CreateFromAssembly()` would also resolve this more specific, but related issue.

**Comment by ManishJayaswal** [comment]
@tmat  moving to 1.2

**Comment by tmat** [comment]
Fixed by #6254

**Comment by loslor** [comment]
Hi,

I've stumbled myself upon this problem. So, I guess it's not solved?

I've used a type in a dynamic Assembly created with Roslyn as daveaglick points out, and then passed a Globals object using this type. It gets a System.NotSupportedException: Can't create a metadata reference to an assembly without location when trying to run the script.

I'm using the DLLs coming in Nuget Microsoft.CodeAnalysis.CSharp.Scripting v2.4.0.

Any lead on what I'm missing (I'm sure I'm missing something) will be much appreciated.

**Comment by rkjellin** [comment]
@loslor Check the pr mentioned above. There is a test named `HostObjectInInMemoryAssembly` showing you how to do it.

**Comment by Nfactor26** [comment]
@rkjellin Test case showing how to achieve this uses a internal sealed class MetadataImagereference which is not accessible. Could someone please provide another example without making use of MetadataImageReference ?

**Comment by MattMinke** [comment]
incase this helps anyone else 

https://stackoverflow.com/a/62223132/2076531
```
using System;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Runtime;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;

namespace Compiler
{
    public static class Compiler
    {
        public static Assembly CompileAssembly(string code, bool compileAsExecutable = false)
        {
            var outputKind = compileAsExecutable ? OutputKind.ConsoleApplication : OutputKind.DynamicallyLinkedLibrary;
            var syntaxTree = CSharpSyntaxTree.ParseText(code);
            var syntaxTrees = new[] {syntaxTree};

            // Requested the reference be fetched
            var runtimeSpecificReference = GetRuntimeSpecificReference();
            var systemAssemblyReference = GetMetadataReference<object>();
            var patchBandAssemblyReference = GetMetadataReference<AssemblyTargetedPatchBandAttribute>();

            // Added the reference to the value here
            var references = new[] {runtimeSpecificReference, systemAssemblyReference, patchBandAssemblyReference};

            var cSharpCompilationOptions = new CSharpCompilationOptions(outputKind);
            var compilation = CSharpCompilation.Create(@"Countdown", syntaxTrees, references, cSharpCompilationOptions);

            using (var assemblyStream = new MemoryStream())
            {
                var emitResult = compilation.Emit(assemblyStream);

                if (emitResult.Success)
                {
                    var assemblyBytes = assemblyStream.ToArray();

                    return Assembly.Load(assemblyBytes);
                }

                var errors = emitResult
                    .Diagnostics
                    .Select(diagnostic => diagnostic.GetMessage())
                    .Select(message => new Exception(message));

                throw new AggregateException(errors);
            }
        }

        private static string GetAssemblyLocation<T>()
        {
            var typeOfT = typeof(T);

            return typeOfT.Assembly.Location;
        }

        private static PortableExecutableReference GetMetadataReference<T>()
        {
            var assemblyLocation = GetAssemblyLocation<T>();

            return MetadataReference.CreateFromFile(assemblyLocation);
        }

        // This function was needed
        private static PortableExecutableReference GetRuntimeSpecificReference()
        {
            var assemblyLocation = GetAssemblyLocation<object>();
            var runtimeDirectory = Path.GetDirectoryName(assemblyLocation);
            var libraryPath = Path.Join(runtimeDirectory, @"netstandard.dll");

            return MetadataReference.CreateFromFile(libraryPath);
        }
    }
}
```
